nab/optimizer.py

```diff
@@ -18,134 +18,303 @@
+    # First anomaly detected inside the window
+    scoreChange += costMatrix["fnWeight"]
+    scoreChange += (computeScaledScoreInsideWindow(dataInfo) *
+                    costMatrix["tpWeight"])
+    # Update the last windows list of detected anomalies
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+  elif dataInfo.idx < dataInfo.lastWindow.detectedAnomalies[0]:
+    # Not first instance of a detected anomaly but earliest in window
+    prevDataInfo = dataInfo.lastWindow.detectedAnomalies[0]
+    scoreChange -= (computeScaledScoreInsideWindow(prevDataInfo) *
+                    costMatrix["tpWeight"])
+
+    scoreChange += (computeScaledScoreInsideWindow(dataInfo) *
+                    costMatrix["tpWeight"])
+
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+  else:
+    # Not first or earliest anomaly in window
+    dataInfo.lastWindow.detectedAnomalies.append(dataInfo)
+    dataInfo.lastWindow.detectedAnomalies.sort(key=lambda di: di.idx)
+
+    # Update countsChange
+    countsChange["fn"] -= 1
+    countsChange["tp"] += 1
+
+  return scoreChange, countsChange
+
+
+
+def _computeThresholdScores(data, numWindows, costMatrix):
+  """Compute the scores for each threshold.
+
+  This is the core of the optimization. It takes the preprocessed data, sorts
+  it by anomaly score (descending), and iteratively lowers the threshold to
+  detect one more row as an anomaly. The threshold is originally set greater
+  than 1.0 and the score is set to the null detector score for the data. Then
+  the threshold is lowered to the anomaly score of each sorted row in turn.
+  Since some thresholds result in multiple rows being detected at once, we only
+  keep the score that corresponds to all of the newly-detected rows being
+  detected.
+
+  We additionally keep track of the counts of true/false positivies/negatives
+  for debugging.
+
+  :param data: a list of _DataInfo instances for all data to be optimized over
+  :param numWindows: the number of anomaly windows in the data which is used to
+      determine the null detector score to start with
+  :param costMatrix: the weights to use in scoring
+  :returns: a list of ThresholdResult instances sorted by score, highest to
+      lowest
   """
-  if not 0 <= threshold <= 1:
-    return float("-inf")
+  # Set up stats
+  counts = {
+    "tn": len(data),
+    "tp": 0,
+    "fp": 0,
+    "fn": 0,
+  }
+  for dataInfo in data:
+    if (dataInfo.lastWindow is not None and
+        dataInfo.idx <= dataInfo.lastWindow.end):
+      counts["tn"] -= 1
+      counts["fn"] += 1
+
+  # Now we iteratively compute the score for every possible threshold
```

Comment:
please add "We start with a threshold >1 such that no detections are made in the dataset, so the score reflects all FNs."